src/application/utils/transaction.ts

```diff
@@ -119,16 +152,39 @@ export async function blindAndSignPset(
     throw new Error('blindPSET error: not fully blinded');
   }
 
-  const signedPset = await signPset(blindedPset, identities);
+  const signedPsetBase64 = await signPset(blindedPset, identities);
+
+  const pset = Psbt.fromBase64(signedPsetBase64);
+  if (!skipSigValidation) {
+    if (!pset.validateSignaturesOfAllInputs(sigValidator)) {
+      throw new Error('PSET is not fully signed');
+    }
+  }
 
-  const decodedPset = decodePset(signedPset);
-  if (!decodedPset.validateSignaturesOfAllInputs()) {
-    throw new Error('PSET is not fully signed');
+  for (let i = 0; i < pset.txInputs.length; i++) {
+    const input = pset.data.inputs[i];
+    // we need to use special finalizer in case of tapscript
+    if (atLeastOne(input.tapLeafScript) && atLeastOne(input.tapScriptSig)) {
+      pset.finalizeInput(i, (_, input) => {
+        return {
+          finalScriptSig: undefined,
+          finalScriptWitness: witnessStackToScriptWitness([
+            ...input.tapScriptSig!.map((s) => s.signature),
+            input.tapLeafScript![0].script,
```

Comment:
how do we ensure the first element is always the correct one? ie. its always spendable?

Reply:
I think we can assume that, at finalize step, signature has been verified. The idea of the finalizer is to use it as "default" to handle the simple cases: simple leaf spendable by marina (basically the $key CHECKSIG script only). This is only used in UI SEND & `marina.sendTransaction` preventing user to coin-select a "complex" leaf.